### PR TITLE
Changing letsencrypt module to use ACME v2 protocol

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Changes the `letsencrypt` module to use the new ACME v2 protocol (as currently specified in https://tools.ietf.org/html/draft-ietf-acme-acme-09), and as supported by the current Let's Encrypt ACME v2 staging endpoint `https://acme-staging-v02.api.letsencrypt.org/directory` (see [announcement](https://letsencrypt.org//2017/06/14/acme-v2-api.html)). This will be necessary to acquire [wildcard certificates from Let's Encrypt](https://letsencrypt.org/2017/07/06/wildcard-certificates-coming-jan-2018.html).

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
`letsencrypt`

##### ANSIBLE VERSION
```
2.4.2.0 (module is from devel branch)
```

##### ADDITIONAL INFORMATION
Current problems:
 1. No more support for current Let's Encrypt ACME v1 endpoint (would be possible to re-integrate that, at a cost of additional complexity);
 2. The certificate chain isn't send by the staging endpoint; no idea whether that's purely a problem of this endpoint though.

### CODE REVIEW

**Comment by felixfontein** [comment]
The missing intermediate cert seems to be a endpoint problem, see letsencrypt/boulder#3291.

Here's a collection of more problems with the staging endpoint: https://community.letsencrypt.org/t/acme-v2-staging-server-known-bugs-issues/49640

This post contains information about the v2 staging endpoint, and points out differences between v1 and v2: https://community.letsencrypt.org/t/staging-endpoint-for-acme-v2/49605

**Comment by felixfontein** [comment]
My current plan is to extend this to support both v1 and v2 endpoints, but I'm not sure whether I will manage this before the end of this week.

**Comment by felixfontein** [comment]
Ok, I think I was able to finish it during today's train ride. The v2 functionality still works, I haven't managed to test the v1 functionality yet.

**Comment by felixfontein** [comment]
TODO:
- ~~add "version check" to see whether the specified endpoint seems to support the specified version~~
- ~~mark `acme_version` as new in 2.5, and mark `agreement` as changed in 2.5~~

File: lib/ansible/modules/web_infrastructure/letsencrypt.py [hunk 1]
```diff
@@ -64,12 +64,22 @@
...
   agreement:
     description:
-      - "URI to a terms of service document you agree to when using the
-         ACME service at C(acme_directory)."
-      - Default is latest gathered from C(acme_directory) URL.
+      - "Boolean indicating whether you agree to the terms of service document."
+      - "ACME servers can require this to be true."
+      - "Note: this setting was changed in version 2.5 to be a boolean, and no
+         longer is the URL of the terms of service document."
+    default: false
```
Comment (resmo) [inline lib/ansible/modules/web_infrastructure/letsencrypt.py]: Changing an argument will break users playbooks, which won't be accepted.

Please create a new argument for this e.g. `term_agreed`.

**Review by resmo — CHANGES_REQUESTED** [review]
Thanks for your effort, please note the comments inline

File: lib/ansible/modules/web_infrastructure/letsencrypt.py (hunk 1 above)
Comment (felixfontein) [reply lib/ansible/modules/web_infrastructure/letsencrypt.py]: Fixed that. I also changed the defaults for `acme_version` and `acme_directory` to `1` resp. the ACME v1 staging endpoint.

Comment (felixfontein) [reply lib/ansible/modules/web_infrastructure/letsencrypt.py]: NB: the current devel letsencrypt module already breaks existing playbooks due to #22074 (default of `fullchain` is `true`, even though documentation says its `false`).

**Comment by felixfontein** [comment]
The new version seems to works fine with the v1 staging endpoint, the v2 staging endpoint, and [Pebble](https://github.com/letsencrypt/Pebble).

**Comment by felixfontein** [comment]
ready_for_review

**Comment by felixfontein** [comment]
I think commit 8c85a93c47380d0ad4b72205b95d6007912628a5 should go, since it does not guarantees that the temp file is actually deleted (`fetch_url` calls `module.fail_json` internally). The solution in #34979 is better suited to solve this problem.
I'll do this somewhen during the next days.

**Comment by felixfontein** [comment]
Ok, that commit is now gone and the conflicts as well. I will test it on the weekend (especially w.r.t. Python 3).

**Comment by felixfontein** [comment]
Fixed a mistake while rebasing. Tests with Python 2 work fine.

**Comment by felixfontein** [comment]
Python 3 works fine, too. There's aparently a bug in the devel branch, though, some tasks included with `include_tasks` weren't executed. I had to "inline" them; with that, everything worked fine.

**Comment by felixfontein** [comment]
I've been thinking a bit of how we should proceed with this PR due to the community module freeze on Feb 7. The final ACME v2 endpoint won't be available until then, so we cannot test this PR against it.
I think we have the following options (I order them from most preferred to least preferred by me):

1. We add a big warning that the ACME v2 endpoint hasn't been properly tested, and merge this somewhen before Feb 7. Since the default is ACME v1, it doesn't break backwards compatibility, but if nothing is substantially wrong, it should also work with ACME v2 (and enable users to get wildcard certificates from Let's Encrypt). Also, by testing against [Pebble](https://github.com/letsencrypt/Pebble), the v2 functionality should be pretty much OK.
2. I split up this PR into two PRs, one with all changes restricted to ACME v1 (which should fix #35172), and one with everything related only to ACME v2.
3. We don't merge this or parts of it at all (and don't fix #35172 at all).

@resmo: what do you think?

**Review by resmo — APPROVED** [review]

File: lib/ansible/modules/web_infrastructure/letsencrypt.py [hunk 2]
```diff
@@ -972,10 +1088,12 @@ def main():
     module = AnsibleModule(
         argument_spec=dict(
             account_key_src=dict(type='path', aliases=['account_key']),
-            account_key_content=dict(type='str'),
+            account_key_content=dict(type='str', no_log=True),
             account_email=dict(required=False, default=None, type='str'),
             acme_directory=dict(required=False, default='https://acme-staging.api.letsencrypt.org/directory', type='str'),
+            acme_version=dict(required=False, default=1, type='int'),
```
Comment (resmo) [inline lib/ansible/modules/web_infrastructure/letsencrypt.py:1094]: I would like to add a choice list here `choices=[1, 2]`. You have already added a choice list in the docs section so I may was your plan to add one here too ;)

**Review by resmo — APPROVED** [review]
As long as we keep backwards compatibility I don't why we should block this. In minor releases we can "fix" v2 functionality as long as we do not need new arguments to be added or break v1 functionality.

I am +1 with an "experimental" warning in the docs

**Comment by felixfontein** [comment]
Good :) I'll add the warnings later.

**Comment by cpu** [comment]
> The certificate chain isn't send by the staging endpoint; no idea whether that's purely a problem of this endpoint though.

We've fixed this in Boulder master. The change should be going out to the staging environment this week.

Thanks for updating this module @felixfontein! I was excited to see the PR :-)

**Comment by resmo** [comment]
argh... the bot accidentally interpreted my +1 as "shipit" sigh

**Comment by resmo** [comment]
anyway, adding that comment, no big deal, we wanted this merged anyway ;)

**Comment by felixfontein** [comment]
@cpu: I saw (I was following your PR), but last time I checked (this morning) it still said "not yet deployed" ;)
@resmo: LOL, sometimes the bot is just too clever ;)

**Pull request merged**